modules/FaceExpressions_module/src/FaceExpressionsRetargeting.cpp

```diff
@@ -0,0 +1,181 @@
+         */
+    if (fvad_set_sample_rate(m_fvadObject, VAD_FREQUENCY))
+    {
+        yError() << "Unsupported input frequency.";
+        return false;
+    }
+
+    m_copiedSound.resize(VAD_SAMPLE_LENGTH);
+
+    yInfo() << "Started";
+
+
+    return true;
+}
+
+double FaceExpressionsRetargetingModule::getPeriod()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    return m_period;
+}
+
+bool FaceExpressionsRetargetingModule::updateModule()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    if (m_audioPort.getPendingReads())
+    {
+        yarp::sig::Sound* inputSound = m_audioPort.read(false);
+
+        if (inputSound->getFrequency() < VAD_FREQUENCY)
+        {
+            yError() << "The frequency needs to be at least " << VAD_FREQUENCY;
+            return false;
+        }
+
+        int subsampling = std::round(inputSound->getFrequency() / (double)VAD_FREQUENCY);
+
+        size_t soundLength = VAD_SAMPLE_LENGTH * subsampling;
+
+        if (soundLength > inputSound->getSamples())
+        {
+            yError() << "The input sound is too short.";
+            return false;
+        }
+
+        for (size_t i = 0; i < VAD_SAMPLE_LENGTH; ++i)
+        {
+            m_copiedSound[i] = inputSound->get(i*subsampling);
+        }
+
+        /*
+         * Calculates a VAD decision for an audio frame.
+         *
+         * `frame` is an array of `length` signed 16-bit samples. Only frames with a
+         * length of 10, 20 or 30 ms are supported, so for example at 8 kHz, `length`
+         * must be either 80, 160 or 240.
+         *
+         * Returns              : 1 - (active voice),
+         *                        0 - (non-active Voice),
+         *                       -1 - (invalid frame length).
+         */
+        m_isTalking = fvad_process(m_fvadObject, m_copiedSound.data(), m_copiedSound.size());
+
+        if (m_isTalking < 0)
+        {
+            yError() << "Invalid frame length.";
+            return false;
+        }
+
+        if (m_isTalking)
+        {
+            m_switchCounter++;
+
+            if (m_switchCounter > m_switchValue)
+            {
+                m_mouthOpen = !m_mouthOpen;
+                m_switchCounter = 0;
+            }
+        }
+        else
+        {
+            m_mouthOpen = false;
+            m_switchCounter = 0;
+        }
+
+        if (m_mouthOpen)
+            m_state="sur";
+        else
+            m_state="neu";
+
+        yarp::os::Bottle cmd, reply;
+        cmd.addVocab(yarp::os::Vocab::encode("set"));
+        cmd.addVocab(yarp::os::Vocab::encode("mou"));
+        cmd.addVocab(yarp::os::Vocab::encode(m_state));
+        m_emotionsOutputPort.write(cmd,reply);
+
+    }
+
+    return true;
+}
+
+bool FaceExpressionsRetargetingModule::close()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    fvad_free(m_fvadObject);
+    m_fvadObject = nullptr;
+    m_audioPort.close();
+    m_emotionsOutputPort.close();
+    yInfo() << "Closing";
+
```

Comment:
@S-Dafarra  Since the process is closing, isn't it cleaner to unlock the object to release the memory? However, the effect should be the same!

Reply:
If you refer to the ``lock``, that is scoped. So as soon as the ``close`` method returns, the ``mutex`` is automatically unlocked. The ``lock`` is necessary here because the ``close`` method might be called while ``updateModule`` is running on a different thread. Since I need to deallocate stuff, it is important I do this only when I am sure no other thread is accessing this data. Hence, it is necessary to lock the mutex.